src/main/java/com/google/codeu/data/Datastore.java

```diff
@@ -146,45 +146,158 @@ public User getUser(String email) {
-  markerEntity.setProperty("lat", marker.getLat());
-  markerEntity.setProperty("lng", marker.getLng());
-  markerEntity.setProperty("content", marker.getContent());
-  datastore.put(markerEntity);
-}
+		Query query = new Query("UserMarker");
+		PreparedQuery results = datastore.prepare(query);
 
-  
-  
-  
-  
+		for (Entity entity : results.asIterable()) {
+			try {
+				double lat = (double) entity.getProperty("lat");
+				double lng = (double) entity.getProperty("lng");
+				String content = (String) entity.getProperty("content");
+
+				UserMarker marker = new UserMarker(lat, lng, content);
+				markers.add(marker);
+			} catch (Exception e) {
+				System.err.println("Error reading marker.");
+				System.err.println(entity.toString());
+				e.printStackTrace();
+			}
+		}
+		return markers;
+	}
+
+	public void storeMarker(UserMarker marker) {
+		Entity markerEntity = new Entity("UserMarker");
+		markerEntity.setProperty("lat", marker.getLat());
+		markerEntity.setProperty("lng", marker.getLng());
+		markerEntity.setProperty("content", marker.getContent());
+		datastore.put(markerEntity);
+	}
+
+	public List<Post> getAllPosts() {
+		Query query = new Query("Post").addSort("timestamp", SortDirection.DESCENDING);
+		PreparedQuery results = datastore.prepare(query);
+
+		List<Post> posts = new ArrayList<>();
+
+		for (Entity entity : results.asIterable()) {
+			try {
+				posts.add(buildPost(entity, false));
+			} catch (Exception e) {
+				System.err.println("Error reading message.");
+				System.err.println(entity.toString());
+				e.printStackTrace();
+			}
+		}
+		return posts;
+	}
+
+	private Post buildPost(Entity entity, boolean isFullPost) {
+		String idString = entity.getKey().getName();
+		UUID postID = UUID.fromString(idString);
+		String creator = (String) entity.getProperty("creator");
+		// String title = (String) entity.getProperty("title");
+		// String content = (String) entity.getProperty("content");
+		String coverImageUrl = (String) entity.getProperty("coverImageUrl");
+		long timestamp = (long) entity.getProperty("timestamp");
+
+		// When we are on index.html and need to show all posts we do not want to
+		// get all this extra information yet because we do not use it on that page.
+		// However when we are on post.html we want to get all the information.
+		//
+		// getAllPosts() uses isFullPost = false
+		// getPost(postID) uses isFullPost = true
+		if (isFullPost) {
+			LocationMarker marker = getLocationMarker(postID);
+			List<ChartDataRow> chartDataRows = getChartDataRows(postID);
+			List<Message> messages = getMessages(postID);
+
+			return new Post(postID, creator, timestamp, coverImageUrl, marker, chartDataRows, messages);
+		}
+
+		return new Post(postID, creator, timestamp, coverImageUrl, null, null, null);
+	}
+
+	private List<ChartDataRow> getChartDataRows(UUID postID) {
+		Query query = new Query("ChartData")
+				.setFilter(new Query.FilterPredicate("postID", FilterOperator.EQUAL, postID.toString()));
```

Comment:
similar to what i commented above. we should figure out if we want the string version here or not. My guess is we do want the string because we store in the datastore as a string.